File: android-kotlin/app/src/main/java/com/example/zipquantum/LinkViewModel.kt
package com.example.zipquantum

import android.app.Application
import android.net.Uri
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

data class LinkUiState(val status: String = "Waiting for a verified link", val destination: String? = null, val working: Boolean = false)

class LinkViewModel(application: Application) : AndroidViewModel(application) {
    private val client = ZipQuantumClient(application)
    private val referrer = InstallReferrerRepository(application)
    private val mutableState = MutableStateFlow(LinkUiState())
    val state: StateFlow<LinkUiState> = mutableState

    fun openAppLink(uri: Uri) = deliver { client.resolve(uri) }

    fun recoverAfterInstall() = viewModelScope.launch {
        val handoff = referrer.deferredHandoff() ?: return@launch
        deliver { client.recover(handoff) }
    }

    private fun deliver(operation: suspend () -> Delivery) = viewModelScope.launch {
        mutableState.value = mutableState.value.copy(working = true)
        runCatching { operation() }.onSuccess { delivery ->
            val destination = delivery.link.destinationUrl ?: delivery.link.url
            mutableState.value = LinkUiState("Route opened: ${delivery.delivery}", destination, false)

            // Move this acknowledgement after your real navigation succeeds.
            val host = delivery.link.host ?: Uri.parse(delivery.link.url).host
            if (delivery.routeAck != null && host != null) client.acknowledgeRouteOpened(delivery.routeAck, host)
        }.onFailure { mutableState.value = LinkUiState(it.message ?: "Delivery failed") }
    }
}

File: ios-swiftui/Sources/LinkViewModel.swift
import Foundation

@MainActor
final class LinkViewModel: ObservableObject {
    @Published private(set) var status = "Waiting for a verified link"
    @Published private(set) var destination: URL?
    @Published private(set) var isWorking = false
    private let client = ZipQuantumClient()

    func openUniversalLink(_ url: URL) { Task { await deliver { try await client.resolve(url) } } }

    func restoreFromPaste(_ value: String) {
        Task {
            do {
                let handoff = try ZQHandoff.parse(value, expectedBundleID: ZQConfiguration.bundleID)
                await deliver { try await client.recover(handoff) }
            } catch { status = error.localizedDescription }
        }
    }

    private func deliver(_ operation: @escaping () async throws -> ZQDelivery) async {
        isWorking = true
        defer { isWorking = false }
        do {
            let delivery = try await operation()
            destination = delivery.link.destinationURL ?? delivery.link.url
            status = "Route opened: \(delivery.delivery)"

            // This line represents the app route becoming visible. In a real app,
            // call acknowledge only after navigation succeeds.
            if let acknowledgement = delivery.routeAcknowledgement,
               let host = delivery.link.host ?? delivery.link.url.host {
                try await client.acknowledgeRouteOpened(acknowledgement, host: host)
            }
        } catch { status = error.localizedDescription }
    }
}
